drivers/bch/bchlib_write.c

```diff
@@ -121,8 +124,45 @@ ssize_t bchlib_write(FAR void *handle, FAR const char *buffer, off_t offset,
       len          -= nbytes;
     }
 
-  /* Then write all of the full sectors following the partial sector
-   * directly from the user buffer.
+  /* Then write all of the full sectors following the partial sector.. */
+#ifdef CONFIG_BCH_FORCE_INDIRECT
+  /* indirectly by using sector buffer.
+   */
+
+  while (len > 0)
+    {
+      /* Read the sector into the sector buffer */
+
+      ret = bchlib_readsector(bch, sector);
+      if (ret < 0)
+        {
+          return ret;
+        }
+
+      /* Copy the data from the user buffer to the sector buffer */
+
+      nbytes = len > bch->sectsize ? bch->sectsize : len;
+      memcpy(bch->buffer, buffer, nbytes);
```

Comment:
bch->buffer needs to be DMA aligned. I presume the MMC device is happy with 8-byte alignment but you'd better check ?

Reply:
bch->buffer alignment can be configured by CONFIG_BCH_BUFFER_ALIGNMENT.

According to techinical reference manual:

> The starting address of the data page must be aligned to the 4 byte boundary (the 2 LSbs set to 0) in 32-bit Addressing mode, and to the 8 byte boundary (the 3 LSbs are set to 0) in 64-bit Addressing mode.

mpfs_dmasendsetup() checks that the buffer is 4-byte aligned and bchlib_flushsector() fails if it's not.